chat-widget/src/common/facades/FacadeChatSDK.ts

```diff
@@ -0,0 +1,284 @@
+        if (this.isAuthenticated) {
+            // if token is not set, or token is already expired , then go to grab a token
+            if (!this.isTokenSet() || this.isTokenExpired() ) {
+                this.token = "";
+                this.expiration = 0;
+                if (this.getAuthToken) {
+                    try {
+                        const ring = await handleAuthentication(this.chatSDK, this.chatConfig, this.getAuthToken);
+                        if (ring.result === true && ring.token) {
+                            await this.setToken(ring.token);
+
+                            TelemetryHelper.logFacadeChatSDKEvent(LogLevel.INFO, {
+                                Event: TelemetryEvent.NewTokenSuccess,
+                                Description: "New Token obtained",
+                                Data: {
+                                    "Token_Expiration": this.expiration
+                                }
+
+                            });
+                            return { result: true, message: "New Token obtained" };
+                        } else {
+
+                            console.error("Failed to get token", ring);
+
+                            TelemetryHelper.logFacadeChatSDKEvent(LogLevel.ERROR, {
+                                Event: TelemetryEvent.NewTokenFailed,
+                                Description: ring.error?.message,
+                                ExceptionDetails: ring.error
+
+                            });
+                            return {
+                                result: false,
+                                message: ring.error?.message || "Failed to get token"
+                            };
+                        }
+
+                    } catch (e: any) {
+
+                        console.error("Unexpected error while getting token", e);
+
+                        TelemetryHelper.logFacadeChatSDKEvent(LogLevel.ERROR, {
+                            Event: TelemetryEvent.NewTokenFailed,
+                            Description: "Unexpected error while getting token",
+                            ExceptionDetails: e
+                        });
+                        return { result: false, message: "Unexpected error while getting token" };
+                    }
+                } else {
+
+                    TelemetryHelper.logFacadeChatSDKEvent(LogLevel.ERROR, {
+                        Event: TelemetryEvent.NewTokenFailed,
+                        Description: "GetAuthToken function is not present",
+                        ExceptionDetails: "Missing function : " + getAuthClientFunction(this.chatConfig)
+                    });
+                }
+            }
+            return { result: true, message: "Token is valid" };
+        }
+
+        return { result: true, message: "Authentication no needed" };
+    }
+
+    private async withTokenRing<T>(functionName: string, fn: () => Promise<T>): Promise<T> {
```

Comment:
Name `withTokenRing` is not clear on the what the intention of this method is

Reply:
you are killing my vibe...